### PR TITLE
Add remote syslog support

### PR DESCRIPTION
**Pull request by nitomartinez**
This change adds:
- rsyslog package
- If the environment variable UDPLOGHOST or TCPLOGHOST is defined, then rsyslog is started inside the container with remote syslog to the $UDPLOGHOST or $TCPLOGHOST. If not defined rsyslog is not started

### CODE REVIEW

**Review by jnovack — CHANGES_REQUESTED** [review]
Don't need to clutter `.dockerignore` and `.gitignore` with your custom ignore files.  Please stick to the official ones.

File: Dockerfile [hunk 1]
```diff
@@ -14,8 +21,15 @@ RUN mkdir -m 0755 -p /var/cache/bind && touch /var/cache/bind/docker-init && cho
...
+VOLUME ["/var/spool/rsyslog"]
+VOLUME ["/etc/rsyslog"]
 VOLUME ["/etc/bind"]
 VOLUME ["/var/cache/bind"]
 
 COPY entrypoint.sh /
+COPY rsyslog.conf /etc/rsyslog.conf
+
+EXPOSE 53
+EXPOSE 53/udp
```
Comment (jnovack) [inline Dockerfile:33]: Make the EXPOSEs a separate pull request.  Two changes in one.

File: README.md [hunk 2]
```diff
@@ -11,12 +12,22 @@ You can change that folder as needed, but make sure you update the "-v" mounts f
...
+
 ## Run BIND Container:
 ```
-docker run --name=dns-slave01 -d --dns=8.8.8.8 --dns=8.8.4.4 \
--p 53:53/udp -p 53:53 \
--v /dns-slave01/etc/bind:/etc/bind \
--v /dns-slave01/var/cache/bind:/var/cache/bind \
--v /dns-slave01/var/log/named:/var/log/named \
-ventz/bind
+docker run --name=dnsserver --hostname=dnsserver --dns=8.8.8.8 --dns=8.8.4.4 -p 53:53/udp -p 53:53 -v /dns/etc/bind:/etc/bind:ro -v /dns/var/cache/bind:/var/cache/bind -v /dns/var/log/named:/var/log/named qindel/bind-syslog
```
Comment (jnovack) [inline README.md:27]: This repo is `ventz/bind` please update.

File: example/etc/bind/named.conf.options [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        category lame-servers { null; };
+        //category cname { null; };
+        // category client { named_debug; };
+        // category resolver { named_debug; };
+        // category unmatched { named_debug; };
+};
+															
+options {
+	directory "/var/cache/bind";
+	version "Qindel Name Server";
```
Comment (jnovack) [inline example/etc/bind/named.conf.options:21]: Not for this repo...

File: example/etc/bind/named.conf.options [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+	// query-source address * port 53;
+
+	// If your ISP provided one or more IP addresses for stable 
+	// nameservers, you probably want to use them as forwarders.  
+	// Uncomment the following block, and insert the addresses replacing 
+	// the all-0's placeholder.
+
+	auth-nxdomain no;    # conform to RFC1035
+	listen-on-v6 { any; };
```
Comment (jnovack) [inline example/etc/bind/named.conf.options:37]: Please still listen on ipv4.

**Review by jnovack — CHANGES_REQUESTED** [review]

**Comment by nitomartinez** [comment]
Thanks for the feedback,

Wil close it for now, and reraise with the modifications

**Comment by jnovack** [comment]
Is rsyslog necessary in every docker container since the update?

https://docs.docker.com/engine/admin/logging/overview/#syslog

**Comment by nitomartinez** [comment]
I think that this is the logging of the container, not the logging of the application.

What I mean is that:

- If the app logs to stdout/stderr then the container will forward it to syslog/file whatever the docker logging driver uses
-  If the app natively logs to syslog or to /dev/log then you will have a way to forward those logs to somewhere outside the container.

there seem to be several approaches to this, but I found starting a syslog inside the container the easiest one...

**Comment by jnovack** [comment]
> there seem to be several approaches to this, but I found starting a syslog inside the container **the easiest one...** (emphasis mine)

That's my concern and why I raise the question.

With the syslog logging engine defined, anything logged to `/dev/stdout` automatically goes where it should given the settings.

It requires you to have type more on the command line each time you start a container, but you should be doing `docker-compose up` to take care of that anyway.

It prevents you from having to install an init or monitor process AND rsyslog on every container.  I noticed you currently aren't using an init process, which means that if your rsyslog daemon fails, you won't know, and your logs are lost.

**Comment by ventz** [comment]
Just seeing this PR/thread -- not sure why github never pinged me via email.

I'll add this merge to the main one, but in a separate branch/tag, since one of the main feedbacks I have received is that people like the simplify of the current image.

That said, central logging is vital in my opinion.

**Comment by ventz** [comment]
@jnovack @nitomartinez Created a new branch ":syslog" -- can you please clean up the PR and I can merge it. Things that need some cleaning up:

1.) Please remove .dockerignore and .gitignore
2.) Dockerfile: I would add the 'rsyslog' from ```RUN apk --update add rsyslog``` on the same line as 'bind'. Same for the mkdir. As you chain, dockerhub builds faster/re-uses layers. (there is also an extra space between ```/var/spool/rsyslog &&  addgroup syslog```
3.) README "run instructions", so that it follows the same syntax I had for the run - one param per line
Also repo name -- see @jnovack's comment - if merging anyway
4.) named.conf.options -> name again

@jnovack @nitomartinez - If you want, I can merge the changes manually and give you credit in the README. (It might be easier/quicker.)

**Comment by ventz** [comment]
I updated with a cleaned up version.

@jnovack Brings up a few very good points there:
1.) As of recently, you can now hook STDOUT and redirect it to a logging engine.
2.) Once you are launching more than one thing - you really should use something like supervisor, which will keep track, restart, notify, etc.

I still think I will create a "syslog" branched version, but I will clean it up a bit.

I'll add another update with some sample zones/examples etc (I think that was a good suggestion).

Closing this for now, but opened an issue to keep track of this.

**Comment by ventz** [comment]
@nitomartinez, @jnovack - Have not forgotten about this either. I know there was a bit of a delay. Still need to clean up a few dir related things/examples/etc, and then i'll focus on adding a syslog branch. I just want to make sure it's against the cleaned up version.

**Comment by jnovack** [comment]
I am against syslog at this point.  This past year more clearly defined the roles of containers and management in the Docker eco-system.  The container should not be managing logging at all, your host/management layer should take care of logging for all containers to a central location.

**Comment by ventz** [comment]
@jnovack I am personally on the same page as you about this. Using the built in logging driver is a much better way to go, and then just making sure all the relevant messages are sent to stdout and stderr. This lets people choose either a straightforward syslog backend or something more fancy like fluentd/cloudwatch/etc. I am glad you chimed in because I was basically going to create a dedicated syslog container and allow it to "attach" itself and "monitor" the DNS logs. But that would have been a separate container in reality.

All that's really needed is something like when you launch the container:
```
-–log-driver syslog –-log-opt syslog-address=udp://a.b.c.d:514
```

I am going to leave a link to the "better option" heer for others who stumble on this:
https://docs.docker.com/engine/admin/logging/overview/

**Pull request closed without merging**